### PR TITLE
[JENKINS-72841] Add `TagsAction` on all relevant stages

### PR DESCRIPTION
**Pull request by timja**
* JENKINS issue(s):
    * [JENKINS-72841](https://issues.jenkins.io/browse/JENKINS-72841)
    * Fixes https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/220
* Description:
    * Previously skipping stages would only be tagged for the first occurrence of the stage. This now checks that the stage is a child of the parent stage. This was only a problem for matrix builds
* Documentation changes:
    * Link to related jenkins.io PR or explanation for why doc change not needed
* Users/aliases to notify:
    * ...

Before:
image

After:

image

see https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/220

### CODE REVIEW

**Review by dwnusbaum — COMMENTED** [review]
I am curious to see what this patch does with a test case like this, with the `when` on the first and then the second branch:

```
pipeline {
    agent none
    stages {
        stage('BuildAndTest') {
          when {
              branch 'testing' // Whatever
          }
          steps { echo 'First copy' }
        }
        stage('BuildAndTest') {
          // try when here too
          steps { echo 'Second copy' }
        }
...
```

Also FWIW, I think this might be a rare case where the `getContext` step could actually make sense.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 1]
```diff
@@ -308,8 +329,19 @@ class Utils {
         }
     }
 
-    static void markStageWithTag(String stageName, String tagName, String tagValue) {
-        List<FlowNode> matched = findStageFlowNodes(stageName)
+    static void markStageWithTag(String stageName, String parentStageName, String tagName, String tagValue) {
```
[the remark this answers is not in the document]
Comment (timja) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy:328]: technically yes and no, its used in ModelInterpreter.groovy

but its groovy so visibility modifiers don't do anything IIUC, but it should not be private

[the remark this answers is not in the document]
Comment (dwnusbaum) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy:328]: Oh sorry, this was supposed to be on the overload that takes `FlowNode` below.

[the remark this answers is not in the document]
Comment (timja) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy:324]: done

[the remark this answers is not in the document]
Comment (timja) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy:324]: can't actually be done unless we rename the function: https://youtrack.jetbrains.com/issue/IDEA-292722/False-positive-Mixing-private-and-public-protected-methods-of-the-same-name

**Comment by timja** [comment]
> I am curious to see what this patch does with a test case like this, with the `when` on the first and then the second branch:

> WorkflowScript: 3: Duplicate stage name: "BuildAndTest" @ line 3, column 5.

This is only an issue with matrix.

I have a more complex example with matrix:

```
pipeline {
  agent any

  stages {
    stage('Hello') {
      steps {
        echo 'Hello World'
      }
    }
    stage('Kubernetes Management Tasks') {
      matrix {
        axes {
          axis {
            name 'K8S_CLUSTER'
            values 'cik8s', 'doks', 'doks-public', 'eks-public', 'privatek8s', 'publick8s'
          }
        } // axes
        stages {
          stage('Prepare Environment'){
            steps {
              echo 'Hello World'
            }
          }
          stage('Helmfile Lint'){
            steps {
              echo 'Hello World'
            }
          } // stage
          stage('Diff on Pull Request'){
            when {
                anyOf {
                    changeRequest()
                    environment name: 'K8S_CLUSTER', value: 'privatek8s'
                }
              
            }
            steps {
              echo 'Hello World'
            }
          } // stage
          stage('Apply'){
            when {
                anyOf {
                    branch 'main'
                    environment name: 'K8S_CLUSTER', value: 'publick8s'
                }
            }
            steps {
              echo 'Hello World'
            }
          } // stage
        } // stages
      } // matrix
    } // stage 'stage('Kubernetes Management Tasks')
  }
}
```

image

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 2]
```diff
@@ -256,6 +250,33 @@ class Utils {
...
+        ForkScanner scanner = new ForkScanner()
+
+        FlowNode stage = null
+        if (parentFlowNode != null) {
+            // find potential heads stopping when parent reached
+            def result = scanner.filteredNodes(execution.currentHeads, [parentFlowNode], CommonUtils.isStageWithOptionalName(stageName))
+
+            result.find {
+                // work back to parent and make sure head is parented to the stage
+                def match = scanner.findFirstMatch(it, { flowNode -> flowNode.getId() == parentFlowNode.getId() })
```
[the remark this answers is not in the document]
Comment (timja) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: done

**Comment by timja** [comment]
Thanks for the review! I think I addressed everything.

**Comment by timja** [comment]
Likely will be replaced by https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/700 as a simpler approach

**Review by joseblas — APPROVED** [review]

**Pull request closed without merging**